### DISCUSSION TITLE
Running multiple FSW instances in a monte-carlo frameowrk

### DISCUSSION THREAD

**Discussion by beavuhs** [nasa/fprime#2799] [Q&A] [answered]
I am looking to set up a software-in-the-loop simulation environment with monte-carlo capability. I already have set up our simulation software to talk to our flight-software, but I am now looking to expand that for monte-carlo simulations. Are there any out-of-the-box capabilities or best practices that Fprime offers to developers for this kind of capability? My initial thought was to run multiples instances of our simulation software and Fprime FSW, with each instance configured to talk to each other over a different UDP port range.

**Comment by Joshua-Anderson** [comment]
I don't think any tooling exists yet, but I think you're going down the right path. If you want to run multiple test cases in parallel on the same computer I'd build a python script wrapper that launches the flight software and a gds instance on a free port, then connects to the GDS using the integration testing python API

**Reply by maxfirefly** [reply]
the FSW `*Topology.cpp` file configures all the different `ByteStreamDriverModel` recv and send ports to specified numbers. These ports correspond to the ports the simulation sends and receives to. The goal is to be able to make these send/recv ports in the `*Topology.cpp` file configurable, so they can change run to run, without having to recompile. The current proposed strategy is leveraging environment variables that get set prior to executing an instance of the FSW, and read in at run time. Is there a more straight forward strategy to be leveraged here?

**Reply by Joshua-Anderson** [reply, marked as answer]
@maxfirefly many fprime projects I've worked with use a main() function that takes the ip/port as a command line argument. Here's an example from the Ref App: https://github.com/nasa/fprime/blob/devel/Ref/Main.cpp#L59-L78

Getting the port number from an environmental variable is a great approach too.

**Reply by maxfirefly** [reply]
thats exactly what we were looking for, thanks so much!
